File: src/main/java/rr/language/RRReferencesSearch.java
package rr.language;

import com.intellij.openapi.application.QueryExecutorBase;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.search.searches.ReferencesSearch;
import com.intellij.util.Processor;
import rr.language.psi.RRBuildingLevelNameDecl;
import rr.language.psi.RRBuildingTreeNameDecl;

import java.util.Collection;

public class RRReferencesSearch extends QueryExecutorBase<PsiReference, ReferencesSearch.SearchParameters> {

    @Override
    public void processQuery(
        ReferencesSearch.SearchParameters queryParameters,
        Processor<? super PsiReference> consumer
    ) {
        PsiElement element = queryParameters.getElementToSearch();

        if (element instanceof RRBuildingTreeNameDecl | element instanceof RRBuildingLevelNameDecl) {
            extendSearchTo(element.getText() + "_desc", element, queryParameters);
            extendSearchTo(element.getText() + "_desc_short", element, queryParameters);
            extendSearchTo(element.getText() + "_name", element, queryParameters);

            Collection<String> factionsAndCulturesNames = RRUtil.findAllFactionsAsStrings(element.getProject());
            factionsAndCulturesNames.addAll(RRUtil.findAllCulturesAsStrings(element.getProject()));
            for (String factionOrCultureName : factionsAndCulturesNames) {
                extendSearchTo(element.getText() + "_" + factionOrCultureName, element, queryParameters);
                extendSearchTo(element.getText() + "_" + factionOrCultureName + "_desc", element, queryParameters);
                extendSearchTo(element.getText() + "_" + factionOrCultureName + "_desc_short", element, queryParameters);
            }
        }
    }

    private void extendSearchTo(String word, PsiElement element, ReferencesSearch.SearchParameters queryParameters) {
        queryParameters.getOptimizer().searchWord(word, queryParameters.getEffectiveSearchScope(), false, element);
    }
}

File: src/main/java/rr/language/RRReferencesSearch.kt
package rr.language

import com.intellij.openapi.application.QueryExecutorBase
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import com.intellij.psi.search.searches.ReferencesSearch
import com.intellij.util.Processor
import rr.language.psi.RRBuildingLevelNameDecl
import rr.language.psi.RRBuildingTreeNameDecl
import rr.language.psi.RRUnitDescrDef

class RRReferencesSearch : QueryExecutorBase<PsiReference?, ReferencesSearch.SearchParameters>(true) {
    override fun processQuery(
        queryParameters: ReferencesSearch.SearchParameters,
        consumer: Processor<in PsiReference?>
    ) {
        val element = queryParameters.elementToSearch

        if (element is RRBuildingTreeNameDecl) {
            extendSearchTo("${element.text}_name", element, queryParameters)
        }

        if (element is RRBuildingLevelNameDecl) {
            extendSearchTo("${element.text}_desc", element, queryParameters)
            extendSearchTo("${element.text}_desc_short", element, queryParameters)

            val factionsAndCulturesNames = RRUtil.findAllFactionsAsStrings(element.project) +
                    RRUtil.findAllCulturesAsStrings(element.project)

            for (factionOrCultureName in factionsAndCulturesNames) {
                extendSearchTo("${element.text}_$factionOrCultureName", element, queryParameters)
                extendSearchTo("${element.text}_${factionOrCultureName}_desc", element, queryParameters)
                extendSearchTo("${element.text}_${factionOrCultureName}_desc_short", element, queryParameters)
            }
        }

        if (element is RRUnitDescrDef) {
            extendSearchTo("${element.text}_descr", element, queryParameters)
            extendSearchTo("${element.text}_descr_short", element, queryParameters)
        }
    }

    private fun extendSearchTo(word: String, element: PsiElement, queryParameters: ReferencesSearch.SearchParameters) {
        queryParameters.optimizer.searchWord(word, queryParameters.effectiveSearchScope, false, element)
    }
}